### DISCUSSION TITLE
Implementation approach for OpenvarioMenu

### DISCUSSION THREAD

**Discussion by Scumi** [XCSoar/XCSoar#1207] [General]
I also started working on the OpenvarioMenu, in parallel to @Blaubart (#1107). I recognized that a lot of stuff can be re-used from the KoboMenu, for example:

- most methods in `System.cpp`
- `WPASupplicant.cpp`
- `WifiDialog.cpp`
- ...and some other stuff

My proposal would be to move those into a more generic folder (like EmbeddedMenu ?) with the device-specific menu files in sub-folders (Kobo, OV), then rename Kobo-specific methods (i.e. `IsKoboWifiOn()` to `IsWifiOn()` in `System.cpp`) and add the Openvario-specific code there.

Is there a way to have a macro for Openvario to make the device specific changes within the same (renamed) method?

For example:

```
bool
IsWifiOn()
{
#ifdef KOBO
  char path[64];
  sprintf(path, "/sys/class/net/%s", GetKoboWifiInterface());
  return Directory::Exists(Path{path});
#elif defined(OPENVARIO)
  return Directory::Exists(Path{"/sys/class/net/wlan0"});
#else
  return false;
#endif
}

```

In the end, just the menu structure and device-specific menu items would remain separated.

What do you think @MaxKellermann ?

**Comment by MaxKellermann** [comment]
There is no `OPENVARIO` macro. The OpenVario is not special - it is a regular XCSoar platform, and it should stay that way.
The Kobo platform is special, lots of things are different here, and we made quite a few shortcuts here; like hard-coding the Wifi interface name (because the Kobo always has exactly one and more cannot be added), but we shouldn't do that on the OpenVario (it has a built-in Wifi interface but more can be added via USB).

I don't know if the WPASupplicant code can be used on the OpenVario; the Kobo has no DBus daemon, and thus XCSoar talks directly to WPASupplicant; but for improved compatibility, we should use DBus on OpenVario. For other projects, I've written C++ bindings for libdbus, which I'll add to XCSoar. These libdbus bindings can then be used to talk to systemd-networkd.

Much of WifiDialog can be used on OpenVario, but it requires some abstractions like how to talk to WPASupplicant and how to configure the IP address (systemd-networkd instead of launching udhcpc manually).

I don't think Kobo/System.cpp can be reused. It hard-codes lots of Kobo things, and these need to be auto-detected on an OpenVario.

**Reply by Scumi** [reply]
Okay, thanks for your insights. True, there is no macro for `OPENVARIO`, but I was curious about the possibility.

I started plugging things together in a similar way as for the KoboMenu and re-using WPASupplicant (among other things) and it worked out of the box. But if auto-detection of certain things is the way to go, then I try to dig into that.

**Reply by MaxKellermann** [reply]
If you have something that works, I'll agree to merge it even if things like interface names are hard-coded for now. Since OpenVarioMenu is only supposed to be used on an OpenVario, making certain assumptions (without OPENVARIO macro check) are okay-ish.
For the long-term, we should clean it up; eventually, OpenVario may switch to a different board, or other people might build similar products, and I want the XCSoar code base to be proper and flexible enough to support everything, but that's not the top-most priority. Right now, I think it's more important to replace this ugly OpenVario menu that is just dozens of shell script glued together...
